Packages/com.unity.inputsystem/InputSystem/Events/IMECompositionEvent.cs

```diff
@@ -2,38 +2,117 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.Runtime.InteropServices;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
 using UnityEngine.InputSystem.Utilities;
 
+//// TODO for v2 remove and replace with just string.
+
 namespace UnityEngine.InputSystem.LowLevel
 {
     /// <summary>
-    /// A specialized event that contains the current IME Composition string, if IME is enabled and active.
-    /// This event contains the entire current string to date, and once a new composition is submitted will send a blank string event.
+    /// Deprecated variant of IME composition event. Please use <see cref="IMECompositionEventVariableSize"/> for replacement.
     /// </summary>
+    [Obsolete("Use IMECompositionEventVariableSize instead.")]
     [StructLayout(LayoutKind.Explicit, Size = InputEvent.kBaseEventSize + sizeof(int) + (sizeof(char) * kIMECharBufferSize))]
-    public struct IMECompositionEvent : IInputEventTypeInfo
+    public unsafe struct IMECompositionEvent : IInputEventTypeInfo
     {
         // These needs to match the native ImeCompositionStringInputEventData settings
-        internal const int kIMECharBufferSize = 64;
+        private const int kIMECharBufferSize = 64;
         public const int Type = 0x494D4553;
 
         [FieldOffset(0)]
         public InputEvent baseEvent;
 
         [FieldOffset(InputEvent.kBaseEventSize)]
-        public IMECompositionString compositionString;
+        private int length;
```

Comment:
What if instead of breaking the API here, we leave IMECompositionString with the same data layout as before, but when the string is greater than kIMECharBufferSize characters long, we (ab)use the first 8 bytes of the fixed buffer to point at a string in some centrally managed storage?

Reply:
We would need some versioning then, which is not optimal. I would really prefer to use a normal managed `string` for such case and not reinvent C++ in C#. Question is more how not to break API here.

Reply:
Yep, using a normal string is the "right" way, but we can't do that and not break the API, so I see the choice as delay this PR until version 2 or make a hack like this that would keep the original data layout. I'm not sure what you mean by versioning? I don't see where we'd need something like that.

Reply:
From my undertraining a managed referenced can't be aliased with unmanaged memory, otherwise GC can't collect it.
So because IMECompositionString is not IDisposable, we can't have a finalizer on it, meaning it would need to point to some other place that stores a string.

In that case lifetime of place storing a string and potential usages of it are uncorrelated, meaning someone could store a copy of IMECompositionString for a long time, and then query it's content.

For such cases a standard solution in C++ is to store a pointer + version of the data, and if handle has different version then a central location, then data is lost.

Reply:
I'm leaning towards IMECompositionString2 and calling it a day, we can clean up when mystical version 2 will come if ever.

Reply:
IMECompositionString2 would be problematic because of Keyboard. I don't think we can do that, unless you'd also have Keyboard.onIMECompositionChangeV2.

Yep, I'm saying let it point to some other place. Actually it wouldn't even have to be a pointer, just an integer lookup into some table of strings. And there's no real issue of versioning. Access would be strictly through the IMECompositionString, and lets just make it IDisposable (we need that anyway because of the issue I mentioned above about building a string from a pointer. Don't think that makes a copy.). You might need to do a swap back or something on disposal and update the index in the affected composition string, but that should be fine.